### DISCUSSION TITLE
Just some general clarifications

### DISCUSSION THREAD

**Discussion by francescodiperna** [phax/peppol-reporting#7] [General]
Good morning, thanks to Philip for this library, very useful. 
I have some general question about the data to be stored for each peppol document sended and received.
Those information are not in SBDH so I need to undertand how to find it.

1 - "m_sEndUserID": is the peppol id of to user where the document is sent for the "sending" direction and the peppol id of the user that receive the message for the "receiving" direction?
2 - in the case of affermative answer to the question at point 1 the field "m_sC1CountryCode" and "m_sC4CountryCode" can be calculated from the peppol id? because I do not know where to retrive this information (we use oxalis for sending and receiving message)
3 -  the field "m_sC3ID" how can be calculated? This information is stored inside the peppol network and is not inside the SBDH

Best regards

**Comment by phax** [comment]
Hello @francescodiperna,
* The EndUser ID is something that you need to know based on the "Know Your Customer" policy. So if you are C2 by contract you need to know C1 and as such you need to know the country code. This is to be taken from master data and NOT from the payload
* As sender you need to know C1 country code from master data, as receiver you get the C1 country code in the SBDH and need to know the country code of C4 from master data.
* As a sender you make an SMP call. The response is signed and contains the certificate. Take the Subject CN of that certificate is the C2 ID.
* As a receiver you get a signed AS4 message including the certificate of the sender. Take the Subject CN of that certificate as the C2 ID.

hth

**Reply by francescodiperna** [reply]
Thanks Philip, about the last point I'm not sure to know how to extract the Subject CN, do you have some examples?

Thanks

**Reply by francescodiperna** [reply]
My SMP use phoss library, can I get those information using some rest endpoints of the SMP?

**Reply by phax** [reply]
For the receiving side, you can see https://github.com/phax/phase4-peppol-standalone/blob/main/src/main/java/com/helger/phase4/peppolstandalone/spi/CustomPeppolIncomingSBDHandlerSPI.java#L56 for the code snippets how to do it.

For the sending side, look at https://github.com/phax/phase4-peppol-standalone/blob/main/src/main/java/com/helger/phase4/peppolstandalone/controller/PeppolSenderController.java#L118 how to integrate the field retrieval inside the builder.

**Reply by francescodiperna** [reply]
Thanks Philip, but I do not use Phase4, I use Oxalis for sending and receiving.
For the receiving document can I use some of your library to parse di delivery evidence (DeliveryNonDeliveryToRecipient) to read the information I need for the report?

Bye

**Reply by francescodiperna** [reply]
Inside the delivery evidence there is the original Envelope, there is stored the certificate of the sender, so I can retrive the  m_sC3ID, is it 
 correct?

**Reply by phax** [reply]
Sorry, I tend to use the overview with all the questions asked ;-) Regarding Oxalis I can recommend #6 - for the rest you need to ask it there. The `DeliveryNonDeliveryToRecipient` object is an Oxalis-specific XML, so no :(
To get the C3 ID you need to get the returned SMP query response. If that is not part of your receipt, you need to either make a secondary SMP lookup, file a request to Oxalis or upgrade your Oxalis version (if it was already changed)

**Comment by francescodiperna** [comment]
> * As sender you need to know C1 country code from master data, as receiver you get the C1 country code in the SBDH and need to know the country code of C4 from master data.

For the receiving documents the C1 isn't in the SDBH for every files, it depend on the version of SBDH.
Bye

**Reply by phax** [reply]
Yes, but from January 1st on, it is mandatory. If you are a receiver, and C1 country code is not contained, you can use "ZZ" if you want to handle the message anyway. See the communication from OpenPeppol on that topic.

**Reply by francescodiperna** [reply]
Thanks
